lib/entry/crossDCsStreamPersistence.js

```diff
@@ -0,0 +1,275 @@
+        
+        logger.debug(`====> [streamPersistenceCheck] starting stream persistence check`);
+        // todo: consult with Guy whether to filter received nodes objects according to status...
+        //entryServerNodeFilter.statusEqual = KalturaLiveStatus.PLAYABLE or KalturaLiveStatus.BROADCASTING;
+        return backendClient.getLiveEntryServerNodes(entryId)
+            .then((serverObjs) => {
+                return _.reduce(serverObjs, (memo, dc) => {
+                    return (dc.status === kalturaTypes.KalturaEntryServerNodeStatus.PLAYABLE) || memo;
+                }, false);
+            })
+            .then((isLive) => {
+                entryIsLive = isLive;
+                return qio.read(this._playlistsData[entryId].path);
+            })
+            .then((content) => {
+                let playlist = content;
+                this._playlistsData[entryId].lastProcessedTime = now;
+                if (newEntry) {
+                    logger.debug(`====> [streamPersistenceCheck] [${entryId}] STOPPED STREAMING expiration time set to [${nowStr}]`);
+                    playlist = playlist.replace(this._expireTimeRegex, `$1${now}`);
+                }
+                if (entryIsLive) {
+                    logger.debug(`====> [streamPersistenceCheck] [${entryId}] is LIVE in other DC, presentation end time set to [${presentationTimeStr}]`);
+                    playlist = playlist.replace(this._presentationEndTimeRegex, `$1${presentationEndTime}`);
+                } else {
+                    ended = true;
+                    logger.debug(`====> [streamPersistenceCheck] [${entryId}] ENDED. Presentation end time set to [${nowStr}]`);
+                    playlist = playlist.replace(this._presentationEndTimeRegex, `$1 ${now}`);
+                }
+                return playlist;
+            })
+            .then((playlist) => {
+                return qio.write(this._playlistsData[entryId].path, playlist);
+            })
+            .then(() => {
+                if (ended) {
+                    return this.removeEntry(entryId, true);
+                } else {
+                    return Q.resolve(false);
+                }
+            })
+            .catch((err) => {
+                logger.error(`persistence check failed for [${entryId}]. error: [${ErrorUtils.error2string(err)}]`);
+                return Q.reject(err);
+            });
+    }
+
+    addEntry(entryId) {
+
+        logger.debug(`====> [addEntry] adding [${entryId}] to persistence list`);
+
+        if (this._entriesToProcess[entryId]) {
+            logger.debug("====> [addEntry][%s] entry is already in stream persistence list, ignoring request", entryId);
+            return Q.resolve();
+        } else {
+            this._entriesToProcess[entryId] = entryId;
+            this._playlistsData[entryId] = {
+                path: path.join(persistenceFormat.getEntryBasePath(entryId), persistenceFormat.getMasterManifestName()),
```

Comment:
why do you keep somethign in memory when you can calcualte it every time you need it? please remove redudany data structures